PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterPlayerScripts/ControlScript/MasterControl/DynamicThumbstick.rbxmx

```diff
@@ -0,0 +1,415 @@
+	<External>null</External>
+	<External>nil</External>
+	<Item class="ModuleScript" referent="RBX1E3E090B922241488CF1527B6A17D7C7">
+		<Properties>
+			<string name="Name">DynamicThumbstick</string>
+			<ProtectedString name="Source"><![CDATA[--[[
+	// FileName: DynamicThumbstick
+	// Version 0.9
+	// Written by: jhelms
+	// Description: Implements dynamic thumbstick controls for touch devices
+--]]
+local Players = game:GetService('Players')
+local UserInputService = game:GetService('UserInputService')
+local GuiService = game:GetService('GuiService')
+local RunService = game:GetService('RunService')
+
+local MasterControl = require(script.Parent)
+
+local Thumbstick = {}
+
+--[[ Script Variables ]]--
+while not Players.LocalPlayer do
+	wait()
+end
+local LocalPlayer = Players.LocalPlayer
+
+local Backpack = LocalPlayer:WaitForChild("Backpack")
+local Tools = {}
+local IsAToolEquipped = false
+
+local IsPortraitMode = true
+
+local RevertAutoJumpEnabledToFalse = false
+
+local ThumbstickFrame = nil
+local ImageCenter = nil
+local ImageUnderThumb = nil
+local ImageRod = nil
+
+local IsFollowStick = false
+local ThumbstickFrame = nil
+local OnMoveTouchEnded = nil		-- defined in Create()
+local OnTouchMovedCn = nil
+local OnTouchEndedCn = nil
+local OnRenderSteppedCn
+local currentMoveVector = Vector3.new(0,0,0)
+
+--[[ Constants ]]--
+
+local IMAGE_OUTER_URL = "rbxasset://textures/ui/Disk.png"
+local IMAGE_STICK_URL = "rbxasset://textures/ui/Disk.png"
+local IMAGE_ROD_URL = "rbxasset://textures/ui/Square.png"
+
+local TOUCH_IS_TAP_TIME_THRESHOLD = 0.5
+local TOUCH_IS_TAP_DISTANCE_THRESHOLD = 25
+
+local FadeInAndOutBackground = true
+local FadeInAndOutMaxCount = nil
+local FadeInAndOutMaxAlpha = 0.35
+local FadeInAndOutHalfDuration = 0.3
+
+--[[ Local functionality ]]--
+
+local function adjustFrameSizeAndPosition()
+	if ThumbstickFrame then
+		if IsPortraitMode then
+			ThumbstickFrame.Size = UDim2.new(1, 0, .4, 0)
+			ThumbstickFrame.Position = UDim2.new(0, 0, 0.6, 0)
+		else
+			ThumbstickFrame.Size = UDim2.new(0.5, 0, 1, 0)
+			ThumbstickFrame.Position = UDim2.new(0, 0, 0, 0)
+		end
+	end
+end
+
+spawn(function()
+	while(true) do
+		--IsPortraitMode = newOrientation == Enum.ScreenOrientation.Portrait
+		wait()
+		local size = workspace.CurrentCamera.ViewportSize
+		local mode = size.X < size.Y
+		if mode ~= IsPortraitMode then
+			IsPortraitMode = mode
+			adjustFrameSizeAndPosition()
+		end
+	end
+end)
+
+local function onWorkspaceChanged(property)
+	if property == 'CurrentCamera' then
+		local newCamera = workspace.CurrentCamera
+		if newCamera then
+			local size = newCamera.ViewportSize
+			IsPortraitMode = size.X < size.Y
+			newCamera.Changed:Connect(function(property)
+				if property == 'ViewportSize' then
+					size = newCamera.ViewportSize
+					IsPortraitMode = size.X < size.Y
+					adjustFrameSizeAndPosition()
+				end
+			end)
+		end
+	end
+end
+
+local function setupTool(tool)
```

Comment:
seems like listenToToolEquip or something might be a more descriptive name for what this does